parlai/agents/style_gen/modules.py

```diff
@@ -0,0 +1,266 @@
+            n_positions=n_positions,
+            activation=opt['activation'],
+            variant=opt['variant'],
+            n_segments=n_segments,
+        )
+
+    def __init__(self, opt, dictionary, num_classes: int, personality_as_label: bool):
+        super().__init__(opt, dictionary)
+        self.classifier_head = nn.Linear(opt['embedding_size'], num_classes)
+        self.personality_as_label = personality_as_label
+
+    def forward(self, *xs):
+        """
+        Get output class logits from the model.
+
+        :param xs:
+            - list of inputs to the encoder/decoder. Elements:
+              - text_vec: (LongTensor[bsz, text seqlen])
+              - label_vec: (LongTensor[bsz, label seqlen])
+                  (Only used if not self.personality_as_label)
+        :return:
+            - the model's predicted per-class scores.
+              (FloatTensor[bsz, len(class_list)])
+        """
+
+        if self.personality_as_label:
+            # All tokens go into the encoder and classification is learned from that.
+            assert len(xs) == 1
+            # Only one input allowed
+            bsz = xs[0].size(0)
+            encoder_states = self.encoder(*xs)
+            inputs = self.START.detach().expand(bsz, 1)
+            # Generate most likely class given start token as input
+            latent, _ = self.decoder(inputs, encoder_states)
+            # latent: [bsz, seqlen, emb_dim]
+            scores = self.classifier_head(latent.squeeze(dim=1))
+        else:
+            # Tokens are split between the encoder and decoder and classification is
+            # learned from both.
+            text_vec, label_vec = xs
+            encoder_states = self.encoder(text_vec)
+            latent, _ = self.decoder(label_vec, encoder_states)
+            # latent: [bsz, seqlen, emb_dim]
+            scores = self.classifier_head(latent.mean(dim=1))
+
+        return scores
+
+
+class BatchWithPersonalities(AttrDict):
+    """
+    Adds a 'personalities' field to the batch in the case where personality information
+    is not encoded in any other field.
+    """
+
+    def __init__(self, personalities=None, **kwargs):
+        super().__init__(personalities=personalities, **kwargs)
+
+
+class TransformerDecoderWithEmbeds(TransformerDecoder):
+    def forward(self, input, encoder_state, embedded_input=None, incr_state=None):
+        """
+        Forward pass with the ability to pass in token-embedded inputs.
+        """
+        # TODO: perhaps reduce the amount of code duplicated from TransformerDecoder.
+        #  This would require modularizing several snippets of code inside
+        #  TransformerDecoder methods.
+
+        encoder_output, encoder_mask = encoder_state
+
+        if input is not None:
+            seq_len = input.size(1)
+            positions = input.new(seq_len).long()
+        else:
+            seq_len = embedded_input.size(1)
+            positions = embedded_input.new(seq_len).long()
```

Comment:
This is because, for the baseline in my style-gen paper, I sometimes needed to pass in "soft" token embeddings (i.e. embeddings weighted over the probability distribution of the next token) instead of the tokens themselves